python/src/iceberg/table/base.py

```diff
@@ -0,0 +1,67 @@
+#  Licensed to the Apache Software Foundation (ASF) under one
+#  or more contributor license agreements.  See the NOTICE file
+#  distributed with this work for additional information
+#  regarding copyright ownership.  The ASF licenses this file
+#  to you under the Apache License, Version 2.0 (the
+#  "License"); you may not use this file except in compliance
+#  with the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing,
+#  software distributed under the License is distributed on an
+#  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+#  KIND, either express or implied.  See the License for the
+#  specific language governing permissions and limitations
+#  under the License.
+
+from __future__ import annotations
+
+from abc import ABC
+from typing import Dict
+
+from attrs import Factory, field, frozen
+
+from iceberg.schema import Schema
+
+
+@frozen(kw_only=True)
```

Comment:
Since we're not using runtime type validation, we can use dataclasses here which is basically a slimmed down version of attrs that's been added to the standard library (might even be by the same author although I'm not 100% sure).
```py
from dataclasses import dataclass

@dataclass(kw_only=True)
class TableSpec:
    ...
```

Reply:
I did start with that initially but thought we would eventually need to add validators for every attribute later on to validate the table spec. That's why went with attrs. I have added a TODO in the code as well to signify the same.